server_environment/models/server_env_mixin.py

```diff
@@ -0,0 +1,405 @@
+        for record in self:
+            for field_name in self._server_env_fields:
+                is_editable_field = self._server_env_is_editable_fieldname(
+                    field_name
+                )
+                is_editable = not record._server_env_has_key_defined(
+                    field_name
+                )
+                record[is_editable_field] = is_editable
+
+    def _server_env_view_set_readonly(self, view_arch):
+        field_xpath = './/field[@name="%s"]'
+        for field in self._server_env_fields:
+            is_editable_field = self._server_env_is_editable_fieldname(field)
+            for elem in view_arch.findall(field_xpath % field):
+                # set env-computed fields to readonly if the configuration
+                # files have a key set for this field
+                elem.set('attrs',
+                         str({'readonly': [(is_editable_field, '=', False)]}))
+            if not view_arch.findall(field_xpath % is_editable_field):
+                # add the _is_editable fields in the view for the 'attrs'
+                # domain
+                view_arch.append(
+                    etree.Element(
+                        'field',
+                        name=is_editable_field,
+                        invisible="1"
+                    )
+                )
+        return view_arch
+
+    def _fields_view_get(self, view_id=None, view_type='form', toolbar=False,
+                         submenu=False):
+        view_data = super()._fields_view_get(
+            view_id=view_id, view_type=view_type,
+            toolbar=toolbar, submenu=submenu
+        )
+        view_arch = etree.fromstring(view_data['arch'].encode('utf-8'))
+        view_arch = self._server_env_view_set_readonly(view_arch)
+        view_data['arch'] = etree.tostring(view_arch, encoding='unicode')
+        return view_data
+
+    def _server_env_default_fieldname(self, base_field_name):
+        """Return the name of the field with default value"""
+        options = self._server_env_fields[base_field_name]
+        if options and options.get('no_default_field'):
+            return ''
+        return '%s_env_default' % (base_field_name,)
+
+    def _server_env_is_editable_fieldname(self, base_field_name):
+        """Return the name of the field for "is editable"
+
+        This is the field used to tell if the env-computed field can
+        be edited.
+        """
+        return '%s_env_is_editable' % (base_field_name,)
+
+    def _server_env_transform_field_to_read_from_env(self, field):
+        """Transform the original field in a computed field"""
+        field.compute = '_compute_server_env'
+
+        inverse_method_name = '_inverse_server_env_%s' % field.name
+        inverse_method = partialmethod(
+            ServerEnvMixin._inverse_server_env, field.name
+        )
+        setattr(ServerEnvMixin, inverse_method_name, inverse_method)
```

Comment:
shouldn't this be set on the model class instead?

Reply:
How? `self.__class__` does not work. You might have a point though

Reply:
`type(self)` does not work?

Reply:
> @guewen you might indeed want to add a couple of tests with _inherit (different model name) and _inherits to check it works properly.

yes, with identical field names in both models, I'll add that

Reply:
Thanks for the hint!

I added tests in 04e54e0, but the thing is that I could not have failing tests with `ServerEnvMixin` instead of `type(self)`. It's probably because that anyway the inverse method is doing the same thing so whether it's defined on ServerEnvMixin or the final model, as long as it's reachable, the result is the same. Anyway I changed it to be more explicit.

Regarding the delegation, I think I don't have the same issues than you, because:
* if you don't override the fields `_inherits` env-computed fields, they are handled by the "parent" model
* if you override some fields in the `_inherits` model, in that case you have to use the mixin again which will take care of the setup of these fields. I had to change these lines

```
        if fieldname not in self._fields:
```
to
```
        if fieldname not in self._fields or self._fields[fieldname].inherited:
```

though